python/mrc/_pymrc/tests/test_json_values.cpp

```diff
@@ -0,0 +1,124 @@
+{
+    py::dict py_dict;
+    py_dict[py::str("test"s)] = py::str("this"s);
+
+    JSONValues j{py_dict};
+    auto result = j.to_python();
+
+    EXPECT_TRUE(result.equal(py_dict));
+    EXPECT_FALSE(result.is(py_dict));  // Ensure we actually serialized the object and not stored it
+}
+
+TEST_F(TestJSONValues, ToPythonRootUnserializable)
+{
+    py::object py_dec = py::module_::import("decimal").attr("Decimal")("1.0");
+
+    JSONValues j{py_dec};
+    auto result = j.to_python();
+
+    EXPECT_TRUE(result.equal(py_dec));
+    EXPECT_TRUE(result.is(py_dec));  // Ensure we stored the object
+}
+
+TEST_F(TestJSONValues, ToPythonSimpleDict)
+{
+    py::object py_dec = py::module_::import("decimal").attr("Decimal")("1.0");
+    py::dict py_dict;
+    py_dict[py::str("test"s)] = py_dec;
+
+    JSONValues j{py_dict};
+    py::dict result = j.to_python();
+
+    EXPECT_TRUE(result.equal(py_dict));
+    EXPECT_FALSE(result.is(py_dict));  // Ensure we actually serialized the dict and not stored it
+
+    py::object result_dec = result["test"];
+    EXPECT_TRUE(result_dec.is(py_dec));  // Ensure we stored the decimal object
+}
+
+TEST_F(TestJSONValues, ToPythonNestedDictUnserializable)
+{
+    // decimal.Decimal is not serializable
+    py::object Decimal = py::module_::import("decimal").attr("Decimal");
+
+    py::object py_dec1 = Decimal("1.1");
+    py::object py_dec2 = Decimal("1.2");
+    py::object py_dec3 = Decimal("1.3");
+
+    std::vector<py::object> py_values = {py::cast(1), py::cast(2), py_dec3, py::cast(4)};
+    py::list py_list                  = py::cast(py_values);
+
+    // Test with object in a nested dict
+    py::dict py_dict("a"_a           = py::dict("b"_a = py::dict("c"_a = py::dict("d"_a = py_dec1))),
+                     "other"_a       = py_dec2,
+                     "nested_list"_a = py_list);
+
+    JSONValues j{py_dict};
+    auto result = j.to_python();
+    EXPECT_TRUE(result.equal(py_dict));
+    EXPECT_FALSE(result.is(py_dict));  // Ensure we actually serialized the object and not stored it
+
+    // Individual Decimal instances shoudl be stored and thus pass an `is` test
+    py::object result_dec1 = result["a"]["b"]["c"]["d"];
+    EXPECT_TRUE(result_dec1.is(py_dec1));
+
+    py::object result_dec2 = result["other"];
+    EXPECT_TRUE(result_dec2.is(py_dec2));
+
+    py::list nested_list   = result["nested_list"];
+    py::object result_dec3 = nested_list[2];
+    EXPECT_TRUE(result_dec3.is(py_dec3));
+}
+
+TEST_F(TestJSONValues, ToPythonList)
+{
+    py::object Decimal = py::module_::import("decimal").attr("Decimal");
+    py::object py_dec  = Decimal("1.1");
+
+    std::vector<py::object> py_values = {py::cast(1), py::cast(2), py_dec, py::cast(4)};
+    py::list py_list                  = py::cast(py_values);
+
+    JSONValues j{py_list};
+    py::list result = j.to_python();
+    EXPECT_TRUE(result.equal(py_list));
+    py::object result_dec = result[2];
+    EXPECT_TRUE(result_dec.is(py_dec));
+}
```

Comment:
It may be worth checking the serialization code with a variety of python objects, like a module, a function, a generator,  etc.. and just make sure there aren't any unexpected gotchas.